File: code-analyzer/src/main/java/com/pranav/analyzer/java/JavacAnalyzer.java
package com.pranav.analyzer.java;

import android.content.Context;
import android.content.SharedPreferences;

import com.pranav.common.util.FileUtil;
import com.pranav.project.mode.JavaProject;

import com.sun.source.util.JavacTask;
import com.sun.tools.javac.api.JavacTool;

import io.github.rosemoe.sora.lang.diagnostic.DiagnosticRegion;

import java.io.File;
import java.io.IOException;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.Locale;

import javax.tools.Diagnostic;
import javax.tools.DiagnosticCollector;
import javax.tools.JavaFileObject;
import javax.tools.SimpleJavaFileObject;
import javax.tools.StandardLocation;

public class JavacAnalyzer {

    private final SharedPreferences prefs;
    private DiagnosticCollector<JavaFileObject> diagnostics = new DiagnosticCollector<>();
    private boolean isFirstUse = true;
    private final String currentFile;
    private final JavaProject project;

    public JavacAnalyzer(Context context, String file, JavaProject project) {
        prefs = context.getSharedPreferences("compiler_settings", Context.MODE_PRIVATE);
        this.project = project;
        currentFile = file;
    }

    public void analyze() throws IOException {
        final var output = new File(project.getBinDirPath(), "classes");
        output.mkdirs();
        final var version = prefs.getString("version", "7");
        final var files = getSourceFiles(new File(project.getSrcDirPath()));

        final var javaFileObjects = new ArrayList<JavaFileObject>();
        for (var file : files) {
        javaFileObjects.add(
                new SimpleJavaFileObject(
                        file.toURI(), JavaFileObject.Kind.SOURCE) {
                    @Override
                    public CharSequence getCharContent(boolean ignoreEncodingErrors)
                            throws IOException {
                        return FileUtil.readFile(file);
                    }
                });
        }

        final var tool = JavacTool.create();

        final var standardJavaFileManager =
                tool.getStandardFileManager(
                        diagnostics, Locale.getDefault(), Charset.defaultCharset());
        standardJavaFileManager.setLocation(
                StandardLocation.PLATFORM_CLASS_PATH, getPlatformClasspath());
        standardJavaFileManager.setLocation(StandardLocation.CLASS_PATH, getClasspath());
        standardJavaFileManager.setLocation(
                StandardLocation.SOURCE_PATH, files);

        final var args = new ArrayList<String>();

        args.add("-proc:none");
        args.add("-source");
        args.add(version);
        args.add("-target");
        args.add(version);
        if (Integer.valueOf(version) >= 9) {
            args.add("--system");
            args.add(FileUtil.getDataDir() + "compiler-modules");
        }

        final var task =
                (JavacTask)
                        tool.getTask(
                                null,
                                standardJavaFileManager,
                                diagnostics,
                                args,
                                null,
                                javaFileObjects);

        task.parse();
        task.analyze();
        standardJavaFileManager.close();
        isFirstUse = false;
    }

    public boolean isFirstRun() {
        return isFirstUse;
    }

    public void reset() {
        diagnostics = new DiagnosticCollector<>();
    }

    public ArrayList<DiagnosticRegion> getDiagnostics() {
        final var problems = new ArrayList<DiagnosticRegion>();
        for (var it : diagnostics.getDiagnostics()) {
            if (it.getSource() == null) continue;
//            if (!it.getCode().startsWith("compiler.err.cant.resolve")) {
//                 problems.add(new DiagnosticRegion((int) it.getStartPosition(), (int) it.getEndPosition(), severity));
//             }
            short severity =
                    (it.getKind() == Diagnostic.Kind.ERROR)
                            ? DiagnosticRegion.SEVERITY_ERROR
                            : DiagnosticRegion.SEVERITY_WARNING;
            problems.add(
                    new DiagnosticRegion(
                            (int) it.getStartPosition(), (int) it.getEndPosition(), severity));
        }
        return problems;
    }

    private ArrayList<File> getClasspath() {
        final var classpath = new ArrayList<File>();
        final var clspath = prefs.getString("classpath", "");

        if (!clspath.isEmpty()) {
            for (var clas : clspath.split(":")) {
                classpath.add(new File(clas));
            }
        }
        classpath.add(project.getBinDir() + "classes");
        return classpath;
    }

    private ArrayList<File> getPlatformClasspath() {
        final var classpath = new ArrayList<File>();
        classpath.add(new File(FileUtil.getClasspathDir(), "android.jar"));
        classpath.add(new File(FileUtil.getClasspathDir(), "core-lambda-stubs.jar"));
        classpath.add(new File(FileUtil.getClasspathDir(), "kotlin-stdlib-1.7.10.jar"));
        return classpath;
    }

    private ArrayList<File> getSourceFiles(File path) {
        final var sourceFiles = new ArrayList<File>();
        final var files = path.listFiles();
        if (files == null) {
            return new ArrayList<File>();
        }
        for (var file : files) {
            if (file.isFile()) {
                if (file.getName().endsWith(".java")) {
                    sourceFiles.add(file);
                }
            } else {
                sourceFiles.addAll(getSourceFiles(file));
            }
        }
        return sourceFiles;
    }
}

File: android-compiler/src/main/java/org/cosmic/ide/analyzer/java/JavacAnalyzer.kt
package org.cosmic.ide.analyzer.java

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import androidx.preference.PreferenceManager
import com.sun.source.util.JavacTask
import com.sun.tools.javac.api.JavacTool
import com.sun.tools.javac.file.JavacFileManager
import io.github.rosemoe.sora.lang.diagnostic.DiagnosticRegion
import io.github.rosemoe.sora.lang.diagnostic.DiagnosticDetail
import org.cosmic.ide.CompilerUtil
import org.cosmic.ide.common.util.FileUtil
import org.cosmic.ide.project.JavaProject
import java.io.File
import java.io.IOException
import java.nio.charset.Charset
import java.util.Locale
import javax.tools.Diagnostic
import javax.tools.DiagnosticCollector
import javax.tools.JavaFileObject
import javax.tools.SimpleJavaFileObject
import javax.tools.StandardLocation

class JavacAnalyzer(
    context: Context,
    private val project: JavaProject
) {

    private val prefs: SharedPreferences
    private var diagnostics = DiagnosticCollector<JavaFileObject>()
    private val tool: JavacTool by lazy { JavacTool.create() }
    private val standardFileManager: JavacFileManager by lazy {
        tool.getStandardFileManager(
            diagnostics, Locale.getDefault(), Charset.defaultCharset()
        )
    }
    private var isFirstUse = true
    private val TAG = "JavacAnalyzer"

    init {
        prefs = PreferenceManager.getDefaultSharedPreferences(context)
        standardFileManager.setLocation(
            StandardLocation.PLATFORM_CLASS_PATH, CompilerUtil.platformClasspath
        )
    }

    @Throws(IOException::class)
    fun analyze() {
        val version = prefs.getString("java_version", "7")
        val files = getSourceFiles(File(project.srcDirPath))

        val javaFileObjects = mutableListOf<JavaFileObject>()
        for (file in files) {
            javaFileObjects.add(
                object : SimpleJavaFileObject(
                    file.toURI(), JavaFileObject.Kind.SOURCE
                ) {
                    override fun getCharContent(ignoreEncodingErrors: Boolean): CharSequence {
                        return FileUtil.readFile(file)
                    }
                })
        }

        with(standardFileManager) {
            setLocation(StandardLocation.CLASS_PATH, getClasspath())
            setLocation(
                StandardLocation.SOURCE_PATH, files
            )
        }

        val args = mutableListOf<String>()

        args.add("-proc:none")
        args.add("-source")
        args.add(version!!)
        args.add("-target")
        args.add(version)

        val task =
            tool.getTask(
                null,
                standardFileManager,
                diagnostics,
                args,
                null,
                javaFileObjects
            ) as JavacTask

        task.parse()
        task.analyze()
        standardFileManager.close()
        isFirstUse = false
    }

    fun isFirstRun() = isFirstUse

    fun reset() {
        diagnostics = DiagnosticCollector<JavaFileObject>()
    }

    fun getDiagnostics(): List<DiagnosticRegion> {
        val diagnostic = diagnostics.diagnostics
        val problems = mutableListOf<DiagnosticRegion>()
        for (it in diagnostic) {
            if (it.source == null) continue
            val severity = if (it.kind == Diagnostic.Kind.ERROR) DiagnosticRegion.SEVERITY_ERROR else DiagnosticRegion.SEVERITY_WARNING
            problems.add(
                DiagnosticRegion(
                    it.startPosition.toInt(), it.endPosition.toInt(), severity, 0, DiagnosticDetail(it.getMessage(Locale.getDefault()))
                )
            )
        }
        return problems
    }

    private fun getClasspath(): List<File> {
        val classpath = mutableListOf<File>()
        classpath.add(File(project.binDirPath, "classes"))
        File(project.libDirPath).walk().forEach {
            if (it.extension.equals("jar")) {
                classpath.add(it)
            }
        }

        return classpath
    }

    private fun getSourceFiles(path: File): List<File> {
        val sourceFiles = mutableListOf<File>()

        path.walk().forEach {
            if (it.extension.equals("java")) {
                sourceFiles.add(it)
            }
        }

        return sourceFiles
    }
}
